enable half for some things, so that "fast path" for multinomial

t-vi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50059 disable multinomial for half
* **#50021 enable half for some things, so that "fast path" for multinomial**

without replacement can be used unconditionally

Next step is to drop the dead code.

t-vi committed:
enable half for some things, so that "fast path" for multinomial
without replacement can be used unconditionally

Next step is to drop the dead code.

[ghstack-poisoned]

ngimel commented:
Multinomial for half datatype on cpu was accidentally added in https://github.com/pytorch/pytorch/pull/29266, and I don't think it makes sense to preserve it. All half operations on cpu are painfully slow. I think it makes sense to just disable multinomial half on cpu rather than enable sorting etc for half.

t-vi committed:
Update on "enable half for some things, so that "fast path" for multinomial"


without replacement can be used unconditionally

Next step is to drop the dead code.

[ghstack-poisoned]

t-vi commented:
Sorry, but I don't really get the gh-stack, and this becomes trival when disabling half, so I'll close the PR and open a plain one.
